### PR TITLE
fix wire swap handling in the creation of a phase poly box [TKET-1653]

### PR DESCRIPTION
**Pull request by cqc-melf**
Fixes: [TKET-1653] No wireswap handling in the creation of a phase poly box #126

I have decided to add this in the creation of the phase poly box to make sure it is not used in other cases. This is probably very inefficient in other cases, the efficient solution is found in the decomposition of the ppb.

### CODE REVIEW

**Review by alexcowtan — APPROVED** [review]
I think that this is fine. I'm not sure that the way of looping to get rid of the qubit permutation is the fastest, but I don't understand the issue well enough to claim otherwise.

**Comment by alexcowtan** [comment]
Could we use Zen's token swapping algorithm for an efficient way of putting in explicit swaps from the implicit qubit permutation?

**Comment by cqc-melf** [comment]
After thinking again a bit longer I still think that this is finding the optimal solution. In our idea you are missing the part, that cycles could overlap, which makes it a bit more difficult. I think that the important point is, that this is only modifying the ppb and not the circuit. The data stored in the ppb is independent from the specific circuit. Maybe there is a way to improve the decomposition of the ppb by using the swap algorithm in the case decomposing the box without aas or other special algorithms. But that requires maybe a bit more work to do that...

File: tket/src/Converters/PhasePoly.cpp [hunk 1]
```diff
@@ -147,18 +147,55 @@ Circuit gray_synth(
...
+    }
+  }
+
+  // replace wireswaps with three CX
+  while (newcirc.has_implicit_wireswaps()) {
+    bool foundswap = false;
+
+    qubit_map_t perm = newcirc.implicit_qubit_permutation();
+    for (const std::pair<const Qubit, Qubit>& pair : perm) {
+      if (pair.first != pair.second) {
```
[the remark this answers is not in the document]
Comment (cqc-melf) [reply tket/src/Converters/PhasePoly.cpp:181]: Yes, that is the case when the wire is not involved in the permutation.

**Comment by alexcowtan** [comment]
> After thinking again a bit longer I still think that this is finding the optimal solution. In our idea you are missing the part, that cycles could overlap, which makes it a bit more difficult.

No, every element in the symmetric group Sn can be expressed as a product of disjoint cycles. (It can also be expressed as a product of potentially overlapping cycles, eg. a product of transpositions.)

**Comment by cqc-melf** [comment]
> No, every element in the symmetric group Sn can be expressed as a product of disjoint cycles. (It can also be expressed as a product of potentially overlapping cycles, eg. a product of transpositions.)

Yes, I forgot that. Thank you!

Are you taking a look if this fixes the other bug or should I do that?

**Pull request merged**